### ISSUE TITLE
Rule Suggestion: Func Style Arrow

### ISSUE THREAD

**Issue by ChristianMurphy** [eslint/eslint#3113]
Requires function be declared using the new arrow syntax over the ES5 function syntax.

This is a stylistic rule, for people that would want to take advantage of the syntactic sugar added in ES6.

Warnings
``` js
let func = function () {return new Promise(function (resolve, reject) { resolve() } )}
func().then(function () {})
```

Not Warnings
``` js
let func = () => {return new Promise((resolve, reject) => { resolve() } )}
func().then(() => {})
```

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
Can you be more specific? Under what circumstances would a function be flagged? Always? (in which case, you're really talking about an extension to `func-style`?) Or just in certain situations?

**Comment by ChristianMurphy** [comment]
It would be flagged always and would do well in `func-style`

**Comment by IanVS** [comment]
My limited understanding is that arrow functions are only syntactic sugar around what used to be accomplished by `self = this;`, but that in the process you give up the true power of `this` style coding, at least according to [YDKJS](https://github.com/getify/You-Dont-Know-JS/blob/master/this%20%26%20object%20prototypes/ch2.md#lexical-this).

Is it therefore useful to create a rule option which would force usage of arrow functions everywhere?

**Comment by ilyavolodin** [comment]
I agree with @IanVS I do not think this rule would be very useful and valid. Arrow functions work completely differently with `this` then regular functions. You will not be able to properly use arrow functions, for example, as properties on the object (at least as far as I understand it). So this rule would produce massive amount of cases where arrows couldn't actually be used.

**Comment by ChristianMurphy** [comment]
I would say it is useful, and *could* produce more valid code.
Lexical scoping was chosen to follow the principle of least astonishment for object oriented style.
Yes it is less flexible than the existing functions, but that is intentional.
Functionally there is nothing that could be acomplished before that can't be accomplished with arrow functions, the syntax is is just different.
In most cases this rule could be enabled and would produce less surprising code (in an OOP sense).

**Comment by nzakas** [comment]
Let's not argue the relative benefits of doing this or not, as that's not how we evaluate rules. I see no harm in adding an arrow function option to the rule. However, func-style only flags functions that are in a declaration, so functions passed to other functions would not be caught by this rule, not would functions on object literals.

So @ChristianMurphy , we really need you to be specific about where your expecting a flag.

**Comment by ChristianMurphy** [comment]
So there are two possible ways that I see this.

1. add an option to func style the will only flag declarations (as you mention)
2. add a new option in the es6 section

something like
``` js
"prefer-arrow": [0, "always", {"except":{ "callback": true}}]
```

where the rule could be enabled always or never.
and exceptions could be made for "declaration", "callback" or "object" where the rule would be ignored.

Warnings for always
``` js
let func = function (callback) {...}
func().then( function () {...} )
let object = {
   func: function () {...}
}
```

Not Warnings for always
``` js
let func = (callback) => {...}
func( () => {...} )
let object = {
   func: () => {...}
}
```

Warnings for never
``` js
let func = (callback) => {...}
func( () => {...} )
let object = {
   func: () => {...}
}
```

Not Warnings for never
``` js
let func = function (callback) {...}
func( function () {...} )
let object = {
   func: function () {...}
}
```

Not Warnings for except declaration
``` js
let func1 = function () {...}
let func2 = () => {...}
```

Not Warnings for except callback
``` js
func( function () {...} )
func( () => {...}
```

Not Warnings for except object
``` js
let obj = {
   es5: function () {...},
   es6: () => {...}
```

Thoughts?

**Comment by nzakas** [comment]
I'm fine with adding an option to `func-style`, that's painless and makes sense.

A `prefer-arrow-function` rule could work, but we couldn't check declarations there because it would overlap with `func-style`. I'd almost prefer a `prefer-arrow-callback` rule instead, that only looks at callbacks. In general, you probably don't want arrow functions attached to objects because you lose `this`.

**Comment by ChristianMurphy** [comment]
:+1: sounds good to me.
should `prefer-arrow-callback` be a new issue?

**Comment by nzakas** [comment]
Yes, that would be helpful.
